_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+```haskell
+tree :: File -> IO ()
+tree file = do
+  dirTree <- buildTree file
+  let formatted = printTree dirTree
+  putStrLn formatted
+```
+But can we go even further?
+
+There seems to be nothing left to extract - we have separated traversal and printing, so what more could we possibly do? But there _is_ in fact a subtle thing. We are walking along the system paths as the trees unfold before us, but we do it in distinct steps: first we take a look around to gather all the possible paths; and then we walk along them.
+
+Let's try to imagine a function like this. First, we need to be able to look around:
+```haskell
+type LookAround a = a -> [a]
+```
+Then, we need to be able to step along the way and construct subtrees - but this is simple recursion. We finally need to be able to construct our node (given the value to store inside the node and the paths to follow):
+```haskell
+type BuildNode t a = a -> [t a] -> t a
+```
+Having these, we can construct our tree builder:
+```haskell
+treeBuilder :: LookAround a -> BuildNode t a -> a -> t a
+treeBuilder lookAround buildNode value = (buildNode value . childNodes . lookAround) value
+  where childNodes = fmap (treeBuilder lookAround buildNode)
+```
+To be more confident that it works as expected, we can, for example, define these:
+```haskell
+data Tree a = Leaf a | Node a [Tree a] deriving Show
+
+children :: LookAround Int
+children 1 = []
+children n = filter (\num -> n `mod` num == 0) [1..n-1]
+
+nodes :: BuildNode Tree Int
+nodes value [] = Leaf value
+nodes value subtrees = Node value subtrees
+```
+
+and then run:
+```haskell
+print (treeBuilder children nodes 12)
+```
+to get a tree in which children of the node are the divisors of the value `v` stored in the node (smaller than `v` itself), `12` being the root.
+
+```
+Node 12 [
+  Leaf 1,
+  Node 2 [Leaf 1],
+  Node 3 [Leaf 1],
+  Node 4 [ 
+    Leaf 1,
+    Node 2 [Leaf 1]
+  ],
+  Node 6 [
+    Leaf 1,
+    Node 2 [Leaf 1],
+    Node 3 [Leaf 1] 
+  ]
+]
+```
+It is neither useful nor spectacular, but gives us some reassurance. The only remaining problem is, how does `IO` fit the puzzle? Obviously, the `treeBuilder`, albeit nice, cannot handle the side effects.
+
+So after another brief moment of hacking, I managed to stitch together an implementation:
+```haskell
+type LookAroundM m a = a -> m [a]
+type BuildNodeM t m a = a -> [t a] -> m (t a)
+
+treeBuilderM :: Monad m => LookAroundM m a -> BuildNodeM t m a -> a -> m (t a)
+treeBuilderM lookAroundM buildNodeM value = lookAroundM value >>= childNodes >>= buildNodeM value
+  where childNodes = traverse (treeBuilderM lookAroundM buildNodeM)
+```
+Now it contains the scary M-word, but it lets us redefine `treeBuilder` as a one-liner:
+```haskell
+treeBuilder lookAround buildNode value = runIdentity (treeBuilderM (return . lookAround) (\a b -> return (buildNode a b)) value)
```

Comment:
Given how far the blogpost goes... I'd mention at the beginning that familarity with Haskell is a must 😅 or, alternatively, consider including Scala-like quasi-equivalents for each snippet (although this will heavily bloat the post 🐡)

Reply:
Yeah, this line is probably unnecessarily complex. It assumes that the monad `m` from `treeBuilderM` is just an identity monad - but there's no way for someone unfamiliar with the language to notice it right away (especially that I don't explain it anywhere).